evals/niah_runner.py

```diff
@@ -0,0 +1,260 @@
+import datetime
+import logging
+import os
+import openai
+import seaborn as sns
+
+from datasets import load_dataset
+import matplotlib.pyplot as plt
+from tqdm import tqdm
+
+from datasets.arrow_dataset import Dataset
+from matplotlib.colors import LinearSegmentedColormap
+from openai.types.beta.assistant import Assistant
+from openai.types.beta.vector_store import VectorStore
+
+logging.basicConfig(level=logging.INFO)
+
+INSTRUCTION_TEMPLATE = """
+                You are a helpful AI bot that answers questions for a user. Keep your response short and direct.
+                You will receive a set of context and a question that will relate to the context.
+                Do not give information outside the document or repeat your findings.
+                If the information is not available in the context respond UNANSWERABLE.
+                """
+
+
+class NIAH_Runner:
+    """
+    A runner to handle executing Needle in a Haystack (NIAH) evals for LeapfrogAI
+
+    This runner assumes LeapfrogAI is already deployed
+    """
+
+    def __init__(
+        self,
+        min_context: int = 0,
+        max_context: int = 32768,
+        dataset: str = "defenseunicorns/LFAI_RAG_niah_v1",
+        model: str = "vllm",
+    ):
+        """Initialize the Assistant with an API key and the path to the text file"""
+
+        self.model = model
+        self.client = openai.OpenAI(
+            base_url=os.getenv("LEAPFROGAI_API_URL"),
+            api_key=os.getenv("LEAPFROGAI_API_KEY"),
+        )
+        logging.info(f"client url: {self.client.base_url}")
+        self.vector_store = self._create_vector_store()
+        self.niah_data = self._load_niah_dataset(dataset, min_context, max_context)
+
+    def evaluate(self) -> None:
+        """Runs the Needle in a Haystack evaluation"""
+
+        retrieval_scores = []
+        response_scores = []
+
+        for row in tqdm(self.niah_data, desc="Evaluating data rows"):
+            logging.info(
+                f"length: {row['context_length']}\n depth: {row['context_depth']}\n secret_code{row['secret_code']}"
+            )
+            # add file to vector_store
+            # TODO: there has to be a more efficient way to add this file
+            with open("context.txt", "wb") as context_file:
+                context_file.write(row["context"].encode("utf-8"))
+            with open("context.txt", "rb") as context_file:
+                vector_store_file = self.client.beta.vector_stores.files.upload(
+                    vector_store_id=self.vector_store.id, file=context_file
+                )
```

Comment:
Try this, it should convert your text directly to bytes, which should be an acceptable type for the `upload` method:
```suggestion
            vector_store_file = self.client.beta.vector_stores.files.upload(
                vector_store_id=self.vector_store.id, file=row["context"].encode('utf-8')
            )
```